google/cloud/aiplatform/models.py

```diff
@@ -2036,3 +2058,138 @@ def list(
+                will be executed in concurrent Future and any downstream object will
+                be immediately returned and synced when the Future has completed.
+
+        Raises:
+            ValueError if model does not support exporting.
+
+            ValueError if invalid arguments or export formats are provided.
+        """
+
+        # Model does not support exporting
+        if not self.supported_export_formats:
+            raise ValueError(f"The model `{self.resource_name}` is not exportable.")
+
+        # No destination provided
+        if not any((artifact_destination, image_destination)):
+            raise ValueError(
+                "Please provide an `artifact_destination` or `image_destination`."
+            )
+
+        export_format_id = export_format_id.lower()
+
+        # Unsupported export type
+        if export_format_id not in self.supported_export_formats:
+            raise ValueError(
+                f"'{export_format_id}' is not a supported export format for this model. "
+                f"Choose one of the following: {self.supported_export_formats}"
+            )
+
+        content_types = gca_model_compat.Model.ExportFormat.ExportableContent
+        supported_content_types = self._supported_export_formats[export_format_id]
+
+        if (
+            artifact_destination
+            and content_types.ARTIFACT not in supported_content_types
+        ):
+            raise ValueError(
+                "This model can not be exported as an artifact. Try exporting as "
+                "a container image by passing the `image_destination` argument."
+            )
+
+        if image_destination and content_types.IMAGE not in supported_content_types:
+            raise ValueError(
+                "This model can not be exported as a container image. Try exporting "
+                "the model artifacts by passing a `artifact_destination` argument."
+            )
+
+        # Construct request payload
+        output_config = gca_model_service_v1.ExportModelRequest.OutputConfig(
+            export_format_id=export_format_id
+        )
+
+        if artifact_destination:
+            output_config.artifact_destination = gca_io_v1.GcsDestination(
+                output_uri_prefix=artifact_destination
+            )
+
+        if image_destination:
+            output_config.image_destination = gca_io_v1.ContainerRegistryDestination(
+                output_uri=image_destination
+            )
+
+        _LOGGER.log_action_start_against_resource("Exporting", "model", self)
+
+        operation_future = self.api_client.export_model(
+            name=self.resource_name, output_config=output_config
+        )
+
+        _LOGGER.log_action_started_against_resource_with_lro(
+            "Export", "model", self.__class__, operation_future
+        )
+
+        # Block before returning
+        operation_future.result()
+
+        _LOGGER.log_action_completed_against_resource("model", "exported", self)
```

Comment:
We should reconsider not returning here. For the export artifact to gcs case, it creates a new directory in the given directory and the user will not have a reference to that directory.

Reply:
Should we return a storage `Blob` to the new dir in the event of a GCS export?

Reply:
I think the usage pattern would be to load the model using the gcs uri after export instead of interacting with the Blob. For example, loading a TF model or passing the uri downstream. I think the main advantage of the blob is that it provides an API to download the file locally but would make the other use cases a bit more laborious.

There are also additional considerations for this API when converted to a Pipeline Component in that a Blob object is not a supported output. 

I'm leaning towards returning ExportModelOperationMetadata.OutputInfo where the key is only present if the output destination is present. 

The main issue with this approach is that our async support only supports returning either a resource noun or None. So we can either remove the sync support for now or you can split the call like this:


```Python
def export(..., sync=True) -> Dict[str, str]:
  ...
  operation_future = self.api_client.export_model(
            name=self.resource_name, output_config=output_config
        )
        
   self._wait_on_export(operation_future, sync=sync)
   
   return convert_to_dict(operation_future.operation.OutputInfo)

@base.optional_sync
def _wait_on_export(self, operation_future, sync=True) -> None:
   operation.result()
```  

Let me know your thoughts.

Reply:
Implemented suggested approach